File: android/app/src/main/java/com/otppoc/android/services/PowService.kt
package com.otppoc.android.services

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.security.MessageDigest

object PowService {

    /**
     * Finds a value where SHA256(nonce + value) has `difficulty` leading zero bits.
     * Runs on Dispatchers.Default to avoid blocking the UI.
     */
    suspend fun solve(nonce: String, difficulty: Int): String = withContext(Dispatchers.Default) {
        val md = MessageDigest.getInstance("SHA-256")
        var counter = 0L
        while (true) {
            val candidate = "$nonce$counter"
            md.reset()
            val hash = md.digest(candidate.toByteArray(Charsets.UTF_8))
            if (hasLeadingZeroBits(hash, difficulty)) {
                return@withContext counter.toString()
            }
            counter++
        }
        @Suppress("UNREACHABLE_CODE")
        throw IllegalStateException("unreachable")
    }

    private fun hasLeadingZeroBits(bytes: ByteArray, count: Int): Boolean {
        var bitsChecked = 0
        for (byte in bytes) {
            if (bitsChecked >= count) return true
            for (bit in 7 downTo 0) {
                if (bitsChecked >= count) return true
                if ((byte.toInt() shr bit) and 1 != 0) {
                    return false
                }
                bitsChecked++
            }
        }
        return bitsChecked >= count
    }
}

File: flutter/lib/services/pow_service.dart
import 'dart:convert';
import 'dart:isolate';
import 'dart:typed_data';

import 'package:crypto/crypto.dart';

class PoWService {
  static Future<String> solve({
    required String nonce,
    required int difficulty,
  }) async {
    return Isolate.run(() => _solveSync(nonce, difficulty));
  }

  static String _solveSync(String nonce, int difficulty) {
    var counter = 0;
    while (true) {
      final candidate = '$nonce$counter';
      final hash = sha256.convert(utf8.encode(candidate));
      final hashBytes = Uint8List.fromList(hash.bytes);

      if (_hasLeadingZeroBits(hashBytes, difficulty)) {
        return counter.toString();
      }
      counter++;
    }
  }

  static bool _hasLeadingZeroBits(Uint8List bytes, int count) {
    var bitsChecked = 0;
    for (final byte in bytes) {
      if (bitsChecked >= count) return true;
      for (var bit = 7; bit >= 0; bit--) {
        if (bitsChecked >= count) return true;
        if ((byte >> bit) & 1 != 0) {
          return false;
        }
        bitsChecked++;
      }
    }
    return bitsChecked >= count;
  }
}
